x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -273,12 +274,16 @@ private RoleReferenceIntersection buildRoleReferencesForCrossClusterAccess() {
         );
         final var innerAuthentication = (Authentication) metadata.get(CROSS_CLUSTER_ACCESS_AUTHENTICATION_KEY);
         final User innerUser = innerAuthentication.getEffectiveSubject().getUser();
-        if (CrossClusterAccessUser.is(innerUser)) {
+        if (innerUser instanceof InternalUser internalUser) {
             assert crossClusterAccessRoleDescriptorsBytes.isEmpty()
                 : "role descriptors bytes list for internal cross cluster access user must be empty";
-            roleReferences.add(
-                new RoleReference.FixedRoleReference(CrossClusterAccessUser.ROLE_DESCRIPTOR, "cross_cluster_access_internal")
-            );
+            final RoleDescriptor internalRoleDescriptor = internalUser.getRemoteAccessRole()
+                .orElseThrow(
+                    () -> new ElasticsearchSecurityException(
+                        "The internal user [" + internalUser + "] is not permitted to perform cross cluster actions"
+                    )
+                );
+            roleReferences.add(new RoleReference.FixedRoleReference(internalRoleDescriptor, "cross_cluster_access_internal"));
```

Comment:
@ywangd Please check that what I've done here looks right to you.

Reply:
Yeah this looks right.